pkg/controller/vsphere/reconciler.go

```diff
@@ -135,12 +140,41 @@ func (r *Reconciler) delete() error {
 // reconcileMachineWithCloudState reconcile machineSpec and status with the latest cloud state
 func (r *Reconciler) reconcileMachineWithCloudState(vm *virtualMachine) error {
 	klog.V(3).Infof("%v: reconciling machine with cloud state", r.machine.GetName())
-	// TODO: reconcile providerID
 	// TODO: reconcile task
+
+	klog.V(3).Infof("%v: reconciling providerID", r.machine.GetName())
+	if err := r.reconcileProviderID(vm); err != nil {
+		return err
+	}
 	klog.V(3).Infof("%v: reconciling network", r.machine.GetName())
 	return r.reconcileNetwork(vm)
 }
 
+func (r *Reconciler) reconcileProviderID(vm *virtualMachine) error {
+	biosUUID := vm.Obj.UUID(vm.Context)
+	if biosUUID == "" {
+		return fmt.Errorf("%v: error getting biosUUID for vm", r.machine.GetName())
+	}
+
+	providerID, err := convertUUIDToProviderID(biosUUID)
+	if err != nil {
+		return err
+	}
+	r.machine.Spec.ProviderID = &providerID
+	return nil
+}
+
+// convertUUIDToProviderID transforms a UUID string into a provider ID.
+// If the supplied UUID is empty or invalid an error is returned.
+// A valid UUID should adhere to the format specified by UUIDPattern.
+func convertUUIDToProviderID(UUID string) (string, error) {
+	pattern := regexp.MustCompile(UUIDPattern)
```

Comment:
I don't really have an objection to this, but we could also just `uuid.Parse()` the UUID and then `parsedUUID.String()` to make sure its in a known format for the provider ID. Might be a little more forgiving, and does away with the big regex.